marahgh/assignment2/BinaryTreeAncestors.h

```diff
@@ -0,0 +1,179 @@
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+using namespace std;
+/*The Binary Tree class,it has no duplicate values */
+template <typename S>
+class BinaryTree {
+  /*Node's Tree class*/
+  template <typename T>
+  class Node {
+   private:
+    T data;
+    Node<T>* left;
+    Node<T>* right;
+    friend class BinaryTree;
+
+   public:
+    /*Node's constructor */
+    Node(T data) : data(data), left(nullptr), right(nullptr) {}
+
+    /*Inserts a node as the right son of this node*/
+    bool insertRight(Node<T>* node) {
+      if (right) {
+        return false;
+      }
+      right = node;
+      return true;
+    }
+
+    /*Inserts a node as the right son of this node*/
+    bool insertLeft(Node<T>* node) {
+      if (left) {
+        return false;
+      }
+      left = node;
+      return true;
+    }
+  };
+
+  /****************************************************************************/
+  /*Binary Tree fields: the root node of the tree*/
+  Node<S>* root;
+
+  /****************************************************************************/
+
+  /*Deleting a binary Tree recursive auxiliary function*/
+  void recursiveDelete(Node<S>* root) {
+    if (!root) {
+      return;
+    }
+    recursiveDelete(root->left);
+    recursiveDelete(root->right);
+    delete root;
+  }
+
+  /*an auxiliary insert recursive function of the insert method.
+  returns true if insertion succeeded, flase otherwise.
+  leftOrRight is string which can be either right or left */
+  bool insertElement(Node<S>* treeRoot, Node<S>* node, S parent,
+                     string leftOrRight) {
+    if (!treeRoot) {
+      return false;
+    }
+    if (treeRoot->data == parent) {
+      if (leftOrRight == "right") {
+        return treeRoot->insertRight(node);
+      }
+      return treeRoot->insertLeft(node);
+    }
+    return (insertElement(treeRoot->left, node, parent, leftOrRight) ||
+            insertElement(treeRoot->right, node, parent, leftOrRight));
+  }
+  /*return true if the given key exists in the given root tree,
+  false otherwise*/
+  bool find(Node<S>* root, S key) {
+    if (!root) {
+      return false;
+    }
+    if (root->data == key) {
+      return true;
+    }
+    return (find(root->left, key)) || (find(root->right, key));
+  }
+
+  /*Auxiliary recursive function of findAncestorsInTree in BinaryTree class*/
+  void findAncestors(Node<S>* root, S& key) {
+    if (!root) {
+      return;
+    }
+    if (root->data == key) {
+      return;
+    }
+    if (find(root, key)) {
+      findAncestors(root->left, key);
+      findAncestors(root->right, key);
+      cout << root->data << endl;

```

Comment:
In general, try for the function you are writing to return the value asked for (in this case, the path, a vector<T>, for example), and have the main function do the printing.